Sources/SwiftFusion/Image/Patch.swift

```diff
@@ -87,18 +132,46 @@ extension Tensor where Scalar == Double {
     let result = ArrayImage(self).patch(at: region).tensor
     return self.shape.count == 2 ? result.reshaped(to: [region.rows, region.cols]) : result
   }
+
+  /// Returns the patch of `self` at `region`, and its Jacobian with respect to the center of the
+  /// `region`.
+  ///
+  /// - Parameters:
+  ///   - self: a tensor of image pixels with shape `[height, width, channelCount]`. `height`
+  ///           and `width` are as defined in `docs/ImageOperations.md`.
+  ///   - region: the center, orientation, and size of the patch to extract.
+  ///
+  /// - Returns:
+  ///   - patch: The region at `region`, cropped from `self`.
+  ///   - jacobian: The derivative of the patch operation with respect to `region`. Has shape
+  ///               `[height, width, channelCount, 3]`. The slices along the last dimension are
+  ///                the directional derivatives along the rotation and the `u`, `v` translation
+  ///                components (in that order).
+  public func patchWithJacobian(at region: OrientedBoundingBox)
+    -> (patch: Tensor<Scalar>, jacobian: Tensor<Scalar>)
+  {
+    precondition(self.shape.count == 3, "image must have shape [height, width, channelCount]")
+    let (patch, jacobian) = ArrayImage(self).patchWithJacobian(at: region)
+    return (
+      patch: patch.tensor,
+      jacobian: Tensor<Double>(stacking: [
```

Comment:
I expect that the value semantics does not magically optimize this because the behavior of `Tensor(stacking:)` is probably very opaque to the compiler.

The reason I'm doing it in two steps is that `Tensor` does not support any efficient way of writing elements into it (raw memory access or otherwise). It is much faster to build an array and then copy it into a tensor than to write elements to the tensor directly.

Of course this is slower than writing directly so we should switch it to writing directly if/when we have a tensor that lets us do that efficiently.

Another possibility might be those precomputed Jacobians we talked about last meeting. It sounds like those would involve a small number of tensor operations that the tensor backend would (hopefully) handle efficiently. The precomputed Jacobians are out of scope of this PR though.